Add sysctl option support when creating service

jtakkala opened the pull request:
Closes #3765

Signed-off-by: Jari Takkala jtakkala@gmail.com

aanand commented:
Thanks for getting this in!

The build will probably fail because of the pinned docker-py version in `requirements.txt`. You'll need to update it to install from a specific Git commit.

jtakkala commented:
@aanand: thanks, do you mean that I should reference the current docker-py version in requirements.txt and include that as part of this commit? Current version in master seems to be 1.10.0-dev.

jtakkala commented:
@aanand: pinned docker-py to a specific commit, but the janky build isn't being scheduled. Tried to force push twice. Can you help?

aanand commented:
retest this please

msidorenko commented:
Any progress with this merge ?

jtakkala commented:
I think this may have been waiting for a docker-py release with docker/docker-py#1145. Now that it's been released a couple of days ago I think this should be ready for a merge. Will need an update to requirements.txt for docker-py version 1.10.0.

jtakkala commented:
@aanand I rebased this on master and updated requirements.txt to point to the 1.10.0 version of docker-py, but build seems to fail. Is this because of some tests failing due to 1.10.0?

jtakkala commented:
Rebased on master and tests passing now.

gonsfx commented:
This is exactly what i need.
This and more namespaced sysctls.

Thanks for your work - hopefully it'll get merged soon.

brandontamm commented:
Looks like this passed recently :) Any expected date when this will make it to a release?

dawidmalina commented:
When this change can be merged?

joshwget commented:
Is there a certain naming convention for plural attributes such as these? It seems odd that it's `sysctls` here but `dns_opt` for #3919.

jtakkala commented:
@joshwget I don't mind if we make it the same as the docker command line options, but it seems that some other config options are plural as well ('labels', 'volumes', etc.). @aanand any comments? Also, any reason we can't merge this in? The longer this sits open the further this diverges from the master branch and will need conflicts to be resolved again.

jtakkala commented:
@joshwget FWIW, [docker-py](https://github.com/docker/docker-py/blob/master/docs/hostconfig.md) calls this option sysctls in my commit there, so to change it now I think would require a change to docker-py to make it consistent.

joshwget commented:
@jtakkala I was mostly curious if `sysctls` and `dns_opt` were the final names for these keys. It seems there are already a few inconsistencies with the current keys then.

Pretty interested in seeing this PR merged as well!

yingdi11 commented:
Is there any samples how can I use sysctl in docker-compose.yml file

EthanLozano commented:
I'm looking forward to see this merge

henryptung commented:
Interest here as well. Are there any more reviews to have? Doesn't seem like anyone's assigned to the PR.

awaterma commented:
+1

nifr commented:
The [corresponding PR](https://github.com/docker/docker-py/pull/1145/commits/dec29e1c10be3fbba239891e3bf47dc6b40ee567) for docker-py has already been merged 3+ months ago. 

Any updates here?

lockie commented:
+1
This would be incredibly helpful; the only other solution to e.g. increase `somaxconn` would be `priveledged: true`, which adds unnecessary security attack surface.

dawidmalina commented:
👍

jtakkala commented:
Not sure why there's been no movement on this one. I will rebase on master and add docs to the new docker.github.io project.

jtakkala commented:
Rebased to master and added docs under docker/docker.github.io#521. @shin- any reason we can't get this merged now?

brandontamm commented:
+1

shin- commented:
Sorry, this completely flew under my radar. Reviewing now.

brandontamm commented:
Is it too late to get this PR into the Release Candidates for upcoming release?!

shin- commented on compose/config/config_schema_v1.json:
```diff
@@ -103,6 +103,7 @@
         "restart": {"type": "string"},
         "security_opt": {"type": "array", "items": {"type": "string"}, "uniqueItems": true},
         "shm_size": {"type": ["number", "string"]},
+        "sysctls": {"$ref": "#/definitions/list_or_dict"},
```
We're not adding anything new to the v1 schema anymore as we want to deprecate it in the future.

shin- commented on compose/config/config_schema_v2.0.json:
```diff
@@ -191,6 +191,7 @@
         "restart": {"type": "string"},
         "security_opt": {"type": "array", "items": {"type": "string"}, "uniqueItems": true},
         "shm_size": {"type": ["number", "string"]},
+        "sysctls": {"$ref": "#/definitions/list_or_dict"},
```
I think the support for the option has been added in API 1.24? In which case, this should be moved to the v2.1 config schema (for API 1.24+). v2.0 will need to stay compatible with API 1.22+

joshwget replied:
Are there any references with more information about v2.1? I haven't been able to find much on the differences between v2.0 and v2.1.

jtakkala replied:
Ok, I've amended the commit and pushed with changes only to v2.1 config schema.

joshwget replied:
Just what I was looking for. Thanks @shin-.

shin- commented:
@brandontamm It is, unfortunately. We feature-freezed a couple weeks ago.

brandontamm commented:
Any idea when this may make it into a release?

shin- commented:
@brandontamm No set date yet, but I'd expect January at the latest.
